### ISSUE TITLE
Preserve order of rownames and colnames in splitAsBumpyMatrix

### ISSUE THREAD

**Issue by LiNk-NY** [LTLA/BumpyMatrix#3]
Hi Aaron, @LTLA 

# TL:DR
 `BumpyMatrix::splitAsBumpyMatrix` should preserve the row and column orders based on the `row` and `column` arguments.

# Report

In light of recent changes to the `SummarizedExperiment` constructor, 
we're getting an error in `SingleCellMultiModal` > `SpatialExperiment` where we are using a
`BumpyMatrix` in the `SummarizedExperiment` constructor function.

The example here is from `SpatialExperiment` / `R/SpatialExperiment.R`:

``` r
n <- 1000; ng <- 50; nc <- 20
# sample xy-coordinates in [0,1]
x <- runif(n)
y <- runif(n)
# assign each molecule to some gene-cell pair
gs <- paste0("gene", seq(ng))
cs <- paste0("cell", seq(nc))
gene <- sample(gs, n, TRUE)
cell <- sample(cs, n, TRUE)
# construct data.frame of molecule coodinates
df <- data.frame(gene, cell, x, y)

# (assure gene & cell are factor so that
# missing observations aren't dropped)
df$gene <- factor(df$gene, gs)
df$cell <- factor(df$cell, cs)

# construct BumpyMatrix
mol <- BumpyMatrix::splitAsBumpyMatrix(
    df[, c("x", "y")],
    row = gs, col = cs)
#> Warning in as.integer(row) + (as.integer(column) - 1L) * nlevels(row): longer
#> object length is not a multiple of shorter object length

# get count matrix
y <- with(df, table(gene, cell))
y <- as.matrix(unclass(y))

# construct SpatialExperiment
SingleCellExperiment::SingleCellExperiment(
    assays = list(counts = y, molecules = mol)
)
#> Error in SummarizedExperiment(...): the rownames and colnames of the supplied assay(s) must be NULL or
#>   identical to those of the constructed SummarizedExperiment object
```

<sup>Created on 2021-08-27 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

<details style="margin-bottom:10px;">

<summary>Session info</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                                      
#>  version  R version 4.1.1 Patched (2021-08-22 r80813)
#>  os       Ubuntu 20.10                               
#>  system   x86_64, linux-gnu                          
#>  ui       X11                                        
#>  language (EN)                                       
#>  collate  en_US.UTF-8                                
#>  ctype    en_US.UTF-8                                
#>  tz       America/New_York                           
#>  date     2021-08-27                                 
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package              * version  date       lib source        
#>  backports              1.2.1    2020-12-09 [1] CRAN (R 4.1.0)
#>  Biobase                2.53.0   2021-05-19 [1] Bioconductor  
#>  BiocGenerics           0.39.2   2021-08-18 [1] Bioconductor  
#>  bitops                 1.0-7    2021-04-24 [1] RSPM (R 4.2.0)
#>  BumpyMatrix            1.1.1    2021-07-04 [1] Bioconductor  
#>  cli                    3.0.1    2021-07-17 [1] RSPM (R 4.1.0)
#>  crayon                 1.4.1    2021-02-08 [1] CRAN (R 4.1.0)
#>  DelayedArray           0.19.1   2021-06-25 [1] Bioconductor  
#>  digest                 0.6.27   2020-10-24 [1] RSPM (R 4.1.0)
#>  ellipsis               0.3.2    2021-04-29 [1] RSPM (R 4.2.0)
#>  evaluate               0.14     2019-05-28 [1] CRAN (R 4.1.0)
#>  fansi                  0.5.0    2021-05-25 [1] RSPM (R 4.1.0)
#>  fs                     1.5.0    2020-07-31 [1] CRAN (R 4.1.0)
#>  GenomeInfoDb           1.29.3   2021-07-01 [1] Bioconductor  
#>  GenomeInfoDbData       1.2.6    2021-05-12 [1] Bioconductor  
#>  GenomicRanges          1.45.0   2021-05-19 [1] Bioconductor  
#>  glue                   1.4.2    2020-08-27 [1] CRAN (R 4.1.0)
#>  highr                  0.9      2021-04-16 [1] RSPM (R 4.2.0)
#>  htmltools              0.5.1.1  2021-01-22 [1] CRAN (R 4.1.0)
#>  IRanges                2.27.2   2021-08-23 [1] Bioconductor  
#>  knitr                  1.33     2021-04-24 [1] RSPM (R 4.2.0)
#>  lattice                0.20-44  2021-05-02 [2] CRAN (R 4.1.1)
#>  lifecycle              1.0.0    2021-02-15 [1] RSPM (R 4.1.0)
#>  magrittr               2.0.1    2020-11-17 [1] RSPM (R 4.1.0)
#>  Matrix                 1.3-4    2021-06-01 [2] CRAN (R 4.1.1)
#>  MatrixGenerics         1.5.3    2021-08-05 [1] Bioconductor  
#>  matrixStats            0.60.1   2021-08-23 [1] RSPM (R 4.1.1)
#>  pillar                 1.6.2    2021-07-29 [1] RSPM (R 4.1.0)
#>  pkgconfig              2.0.3    2019-09-22 [1] CRAN (R 4.1.0)
#>  purrr                  0.3.4    2020-04-17 [1] CRAN (R 4.1.0)
#>  RCurl                  1.98-1.4 2021-08-17 [1] RSPM (R 4.1.0)
#>  reprex                 2.0.1    2021-08-05 [1] RSPM (R 4.1.0)
#>  rlang                  0.4.11   2021-04-30 [1] RSPM (R 4.2.0)
#>  rmarkdown              2.10     2021-08-06 [1] RSPM (R 4.1.0)
#>  S4Vectors              0.31.1   2021-08-18 [1] Bioconductor  
#>  sessioninfo            1.1.1    2018-11-05 [1] CRAN (R 4.1.0)
#>  SingleCellExperiment   1.15.2   2021-08-27 [1] Bioconductor  
#>  stringi                1.7.3    2021-07-16 [1] RSPM (R 4.1.0)
#>  stringr                1.4.0    2019-02-10 [1] CRAN (R 4.1.0)
#>  styler                 1.5.1    2021-07-13 [1] RSPM (R 4.1.0)
#>  SummarizedExperiment   1.23.3   2021-08-24 [1] Bioconductor  
#>  tibble                 3.1.3    2021-07-23 [1] RSPM (R 4.1.0)
#>  utf8                   1.2.2    2021-07-24 [1] RSPM (R 4.1.0)
#>  vctrs                  0.3.8    2021-04-29 [1] RSPM (R 4.2.0)
#>  withr                  2.4.2    2021-04-18 [1] RSPM (R 4.2.0)
#>  xfun                   0.25     2021-08-06 [1] RSPM (R 4.1.0)
#>  XVector                0.33.0   2021-05-19 [1] Bioconductor  
#>  yaml                   2.2.1    2020-02-01 [1] CRAN (R 4.1.0)
#>  zlibbioc               1.39.0   2021-05-19 [1] Bioconductor  
#> 
#> [1] /media/mr148/1D24A0EA4286043C1/bioc-devel
#> [2] /home/mr148/src/svn/r-release/R/lib/R/library
```

</details>

The issue is that `BumpyMatrix::splitAsBumpyMatrix` should preserve the row and column orders based on the `row` and `column` arguments.
# Unit test
Perhaps there should be a unit test such as:
```
## unit test
identical(gs, rownames(mol))
identical(cs, colnames(mol))
```

# Patch
Here is what the patch would look like:

```diff
diff --git a/R/splitAsBumpyMatrix.R b/R/splitAsBumpyMatrix.R
index 077b9d1..02c89a7 100644
--- a/R/splitAsBumpyMatrix.R
+++ b/R/splitAsBumpyMatrix.R
@@ -44,8 +44,8 @@
 #' @export
 #' @importFrom Matrix sparseMatrix
 splitAsBumpyMatrix <- function(x, row, column, sparse=FALSE) {
-    row <- as.factor(row)
-    column <- as.factor(column)
+    row <- factor(row, levels = row)
+    column <- factor(column, levels = column)
 
     # TODO: avoid running into the .Machine$integer.max limit. 
     i <- as.integer(row) + (as.integer(column) - 1L) * nlevels(row)
```

Best regards,
Marcel

**Comment by LTLA** [comment]
That's not how `splitAsBumpyMatrix` works. There is no reason that the rows and columns of the output should be in the same order as entries in `row` and `column`. In fact, `row` and `column` don't even have a concept of "order"; they are parallel to `x`, rather than being parallel to their respective dimensions. `gs` and `cs` should be of length equal to `nrow(df)`, hence the warning.

Regardless, the proposed patch is not ideal as it would eliminate any existing factor ordering in `row` and `column`. If you want a specific (non-alphanumeric) ordering, the better approach is to supply factors instead.

**Comment by LiNk-NY** [comment]
Thanks!
